File: py/aoc2022/day2.py
"""
Day 2: Rock Paper Scissors
"""


def _parse(lines):
    return (
        (ord(line[0]) - ord("A") + 1, ord(line[2]) - ord("X") + 1) for line in lines
    )


def _score(left, right):
    return (1 + right - left) % 3 * 3 + right


def part1(lines):
    """
    >>> part1(["A Y", "B X", "C Z"])
    15
    """
    return sum(_score(left, right) for left, right in _parse(lines))


def part2(lines):
    """
    >>> part2(["A Y", "B X", "C Z"])
    12
    """
    return sum(_score(left, 1 + (left + right) % 3) for left, right in _parse(lines))


parts = (part1, part2)

File: hs/src/Day2.hs
{-|
Module:         Day2
Description:    <https://adventofcode.com/2022/day/2 Day 2: Rock Paper Scissors>
-}
{-# LANGUAGE OverloadedStrings #-}
module Day2 (day2a, day2b) where

import Data.Char (ord)
import Data.Text (Text)
import qualified Data.Text as T (head, lines, words)

parse :: Text -> (Int, Int)
parse line = (parse1 lhs, parse1 rhs) where
    [lhs, rhs] = T.words line
    parse1 = succ . (`mod` 3) . (`mod` 23) . pred . ord . T.head

score :: Int -> Int -> Int
score lhs rhs = (1 + rhs - lhs) `mod` 3 * 3 + rhs

day2a :: Text -> Int
day2a input = sum $ uncurry score . parse <$> T.lines input

day2b :: Text -> Int
day2b input = sum [score lhs $ 1 + (lhs + rhs) `mod` 3 | (lhs, rhs) <- parse <$> T.lines input]
